client/rest/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -91,231 +87,138 @@
+     * before giving up and failing the request.
+     */
+    private static final int MAX_NEXT_NODES_ATTEMPTS = 10;
+
     private final CloseableHttpAsyncClient client;
     // We don't rely on default headers supported by HttpAsyncClient as those cannot be replaced.
     // These are package private for tests.
     final List<Header> defaultHeaders;
     private final long maxRetryTimeoutMillis;
     private final String pathPrefix;
-    private final AtomicInteger lastHostIndex = new AtomicInteger(0);
-    private volatile HostTuple<Set<HttpHost>> hostTuple;
+    private final AtomicInteger lastNodeIndex = new AtomicInteger(0);
     private final ConcurrentMap<HttpHost, DeadHostState> blacklist = new ConcurrentHashMap<>();
     private final FailureListener failureListener;
+    private volatile NodeTuple<List<Node>> nodeTuple;
 
     RestClient(CloseableHttpAsyncClient client, long maxRetryTimeoutMillis, Header[] defaultHeaders,
-               HttpHost[] hosts, String pathPrefix, FailureListener failureListener) {
+               Node[] nodes, String pathPrefix, FailureListener failureListener) {
         this.client = client;
         this.maxRetryTimeoutMillis = maxRetryTimeoutMillis;
         this.defaultHeaders = Collections.unmodifiableList(Arrays.asList(defaultHeaders));
         this.failureListener = failureListener;
         this.pathPrefix = pathPrefix;
-        setHosts(hosts);
+        setNodes(nodes);
     }
 
     /**
      * Returns a new {@link RestClientBuilder} to help with {@link RestClient} creation.
-     * Creates a new builder instance and sets the hosts that the client will send requests to.
+     * Creates a new builder instance and sets the nodes that the client will send requests to.
+     * @see Node#Node(HttpHost)
      */
     public static RestClientBuilder builder(HttpHost... hosts) {
-        return new RestClientBuilder(hosts);
+        return builder(hostsToNodes(hosts));
     }
 
     /**
-     * Replaces the hosts that the client communicates with.
-     * @see HttpHost
+     * Returns a new {@link RestClientBuilder} to help with {@link RestClient} creation.
+     * Creates a new builder instance and sets the nodes that the client will send requests to.
      */
-    public synchronized void setHosts(HttpHost... hosts) {
-        if (hosts == null || hosts.length == 0) {
-            throw new IllegalArgumentException("hosts must not be null nor empty");
-        }
-        Set<HttpHost> httpHosts = new HashSet<>();
-        AuthCache authCache = new BasicAuthCache();
-        for (HttpHost host : hosts) {
-            Objects.requireNonNull(host, "host cannot be null");
-            httpHosts.add(host);
-            authCache.put(host, new BasicScheme());
-        }
-        this.hostTuple = new HostTuple<>(Collections.unmodifiableSet(httpHosts), authCache);
-        this.blacklist.clear();
+    public static RestClientBuilder builder(Node... nodes) {
```

Comment:
could we clarify in the docs why one would use this over the other one and what a Node is for compared to a host?

Reply:
Sure!